flow/layers/clip_path_layer.cc

```diff
@@ -15,6 +15,22 @@ ClipPathLayer::ClipPathLayer(const SkPath& clip_path, Clip clip_behavior)
   FML_DCHECK(clip_behavior != Clip::none);
 }
 
+void ClipPathLayer::Diff(DiffContext* context, const Layer* old_layer) {
+  auto subtree = context->BeginSubtree();
+  auto* prev = reinterpret_cast<const ClipPathLayer*>(old_layer);
+  if (!context->IsSubtreeDirty()) {
+    assert(prev);
+    if (clip_behavior_ != prev->clip_behavior_ ||
+        clip_path_ != prev->clip_path_) {
+      context->MarkSubtreeDirty(old_layer->paint_region());
+    }
+  }
+  if (context->PushCullRect(clip_path_.getBounds())) {
+    DiffChildren(context, prev);
+  }
+  set_paint_region(context->CurrentSubtreeRegion());
+}
+
```

Comment:
The pattern in line 19-31 seems to be repetitive in many layers with minor modifications. I wonder if it's cleaner to distill the common skeleton like
```
void Layer::Diff(DiffContext* context, const Layer* old_layer) {
  auto subtree = context->BeginSubtree();
  if (!context->IsSubtreeDirty()) {
    assert(old_layer);
    // Layer subclass can implement IsDirtyComparedTo differently. By default,
    // it can always return true.
    if (IsDirtyComparedTo(old_layer)) {
      context->MarkSubtreeDirty(old_layer->paint_region());
    }
  }

  // Layer subclass can implement GetTransform differently. By default, it will
  // be identity except for TransformLayer and other layers that add offsets.
  if (!GetTransform().isIdentity()) {
    context->PushTransform(GetTransform());
  }

  // Layer subclass can implement GetDirectPaintRegion differently. The child 
  // layer's paint region is not considered in GetDirectPaintRegion. So it will
  // be empty by default except for layers like PictureLayer, TextureLayer,
  // PhysicalShapeLayer, ...
  context->AddPaintRegion(GetDirectPaintRegion());

  // Layer subclass can implement GetPreFilterReadbackRegion differently, By 
  // default, it will be empty except for BackdropFilterLayer.
  context->AddReadbackRegion(GetPreFilterReadbackRegion(context));

  // Layer subclass can implement GetCullRect differently. Infinite cull rect is
  // returned if this layer doesn't have a cull rect. By default, it may
  // always return an infinite cull rect (for non clip layers).
  SkRect cull_rect = GetCullRect();
  if (!cull_rect.isFinite() || context->PushCullRect(cull_rect)) {
    DiffChildren(context, prev);
  }

  // Layer subclass can implement GetPostFilterReadbackRegion differently, By 
  // default, it will be empty except for ImageFilterLayer.
  context->AddReadbackRegion(GetPostFilterReadbackRegion(context));

  set_paint_region(context->CurrentSubtreeRegion());
}
```

Reply:
You could do this, but I wanted to avoid making bunch of virtual methods that would only be used during Diff. You could make similar point about preroll (why don't class just implement GetTransform / GetCullRect, etc. It feels a bit uneven to do it like this for Diff, but not for preroll.

There is some repetition in current approach, but it's also relatively easy to understand.